### PR TITLE
Support new circuit

### PR DESCRIPTION
**Pull request by qzhodl**
- This PR addresses https://github.com/ethstorage/storage-contracts-v1/issues/20. 
 - EthStorageContract supports the old circuit. EthStorageContract2 inherits EthStorageContract, and supports the new circuit.
 - TODO: need to fix the UT

### CODE REVIEW

File: contracts/EthStorageContract.sol [hunk 1]
```diff
@@ -139,25 +138,17 @@ contract EthStorageContract is StorageContract, Decoder {
...
         PhyAddr memory kvInfo = kvMap[idxMap[kvIdx]];
-        (Proof memory decodeProof, uint256 mask, bytes memory peInput) = abi.decode(proof, (Proof, uint256, bytes));
-
-        // BLOB decoding check
-        if (
-            !decodeSample(decodeProof, uint256(keccak256(abi.encode(kvInfo.hash, miner, kvIdx))), sampleIdxInKv, mask)
-        ) {
-            return false;
-        }
+        (uint256 mask, bytes memory peInput) = abi.decode(proof, (uint256, bytes));
```
Comment (syntrust) [inline contracts/EthStorageContract.sol:148]: Could it be better to decode earlier, i.e., moving decode to the loop of  `_verifySamples`, for returning mask is not like a result `checkInclusive` should have.

File: contracts/EthStorageContract.sol [hunk 2]
```diff
@@ -101,22 +101,21 @@ contract EthStorageContract is StorageContract, Decoder {
...
-        uint256[] memory input = new uint256[](3);
-        // TODO: simple hash to curve mapping
-        input[0] = encodingKey % modulusBn254;
-        input[1] = xBn254;
-        input[2] = mask;
-        return (verifyDecoding(input, proof) == 0);
+        uint256[] memory masks,
+        bytes calldata decodeProof
+    ) public view virtual override returns (bool) {        
+        require(masks.length == 2, "invalid mask length");
```
Comment (syntrust) [inline contracts/EthStorageContract.sol:107]: Should we use `randomChecks` to replace 2?

File: contracts/StorageContract.sol [hunk 3]
```diff
@@ -132,28 +139,29 @@ abstract contract StorageContract is DecentralizedKV {
...
-            uint256 kvIdx = sampleIdx >> sampleLenBits;
+            uint256 sampleIdx = uint256(hash0) % rows + (startShardId << (shardEntryBits + sampleLenBits));
             uint256 sampleIdxInKv = sampleIdx % (1 << sampleLenBits);
 
-            require(
-                decodeAndCheckInclusive(kvIdx, sampleIdxInKv, miner, encodedSamples[i], inclusiveProofs[i]),
-                "invalid samples"
-            );
+            (bool success, uint256 mask) = checkInclusive(sampleIdx >> sampleLenBits, sampleIdxInKv, encodedSamples[i], inclusiveProofs[i]);
+            require(success, "invalid samples");
```
Comment (syntrust) [inline contracts/StorageContract.sol:157]: msg could be more precise like "not inclusive"?

File: contracts/StorageContract.sol [hunk 4]
```diff
@@ -222,7 +231,7 @@ abstract contract StorageContract is DecentralizedKV {
 
         // Check if the data matches the hash in metadata and obtain the solution hash.
         bytes32 hash0 = keccak256(abi.encode(miner, bh, nonce));
-        hash0 = _verifySamples(shardId, hash0, miner, encodedSamples, proofs);
+        hash0 = _verifySamples(shardId, hash0, encodedSamples, inclusiveProofs, decodeProof);
```
Comment (syntrust) [inline contracts/StorageContract.sol:234]: Does it make more sense to encode masks with decodeProof other than inclusiveProofs? Or as a separate parameter?

**Review by qizhou — APPROVED** [review]
Looks great!

**Review by syntrust — APPROVED** [review]

File: contracts/EthStorageContract2.sol [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            PhyAddr memory kvInfo = kvMap[idxMap[kvIdx]];
+
+            require(
+                checkInclusive(kvInfo.hash, sampleIdxInKv, masks[i] ^ uint256(encodedSamples[i]), inclusiveProofs[i]),
+                "invalid samples"
+            );
+            hash0 = keccak256(abi.encode(hash0, encodedSamples[i]));
+        }
+
+        require(decodeSample(masks, decodeProof[0]), "decode failed");
```
Comment (qizhou) [inline contracts/EthStorageContract2.sol:57]: I think this is wrong somewhere
- miner is not used in the decoding and proving (as part of the encodingKey).

**Pull request merged**